velocity/src/main/java/co/aikar/commands/ACFVelocityUtil.java

```diff
@@ -0,0 +1,90 @@
+package co.aikar.commands;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Sets;
+import com.velocitypowered.api.command.CommandSource;
+import com.velocitypowered.api.proxy.Player;
+import com.velocitypowered.api.proxy.ProxyServer;
+
+import net.kyori.text.TextComponent;
+import net.kyori.text.serializer.ComponentSerializers;
+
+public class ACFVelocityUtil {
+
+    @SuppressWarnings("deprecation")
+    public static TextComponent color(String message) {
+        return ComponentSerializers.LEGACY.deserialize(message);
+    }
+
+    public static Player findPlayerSmart(ProxyServer server, CommandIssuer issuer, String search) {
+        CommandSource requester = issuer.getIssuer();
+        String name = ACFUtil.replace(search, ":confirm", "");
+        if (name.length() < 3) {
+            issuer.sendError(MinecraftMessageKeys.USERNAME_TOO_SHORT);
+            return null;
+        }
+        if (!isValidName(name)) {
+            issuer.sendError(MinecraftMessageKeys.IS_NOT_A_VALID_NAME, "{name}", name);
+            return null;
+        }
+
+        List<Player> matches = new ArrayList<>(matchPlayer(server, name));
+
+        if (matches.size() > 1) {
+            String allMatches = matches.stream().map(Player::getUsername).collect(Collectors.joining(", "));
+            issuer.sendError(MinecraftMessageKeys.MULTIPLE_PLAYERS_MATCH, "{search}", name, "{all}", allMatches);
+            return null;
+        }
+
+        if (matches.isEmpty()) {
+            issuer.sendError(MinecraftMessageKeys.NO_PLAYER_FOUND_SERVER, "{search}", name);
+            return null;
+        }
+
+        return matches.get(0);
+    }
+
+    /*
+     * Original code written by md_5
+     * 
+     * Modified to work with Velocity by Crypnotic
+     */
+    private static Collection<Player> matchPlayer(ProxyServer server, final String partialName) {
+        Preconditions.checkNotNull(partialName, "partialName");
+
+        Optional<Player> exactMatch = server.getPlayer(partialName);
+        if (exactMatch != null) {
+            return Collections.singleton(exactMatch.get());
+        }
+
+        return Sets.newHashSet(Iterables.filter(server.getAllPlayers(), new Predicate<Player>() {
```

Comment:
Remove the unneeded usage of Guava. We've got Java 8 here:

```java
return server.getAllPlayers().stream()
  .filter(player -> player.getUsername().toLowerCase(Locale.US).startsWith(partialName.toLowerCase(Locale.US)))
  .collect(Collectors.toList());
```

This could be even more efficient with `String#regionMatches()`, but I'll leave that as an exercise to you.

Reply:
Alternatively, I could just add `matchPlayer()` to the Velocity API.

Reply:
I think other people could find a nice use `matchPlayer()`